Construct EVM transactions inside Executor

felixfaisal opened the pull request:
So this PR as title attemps to construct transactions inside the executor itself using aggregators such as okx, 1inch and kyberswap. 

Majority of the code has been referenced from the pumpx backend, I've not integrated this with the current cross-chain execution and requesting signer, because i haven't been able to do a test with the aggregators, and I've created an issue for it P-1590, I've created integrations tests under examples/ to test them. 

This PR also does not cover four meme aggregator, it'll be covered under a separate issue P-1591

felixfaisal committed:
init: evm aggregators client

felixfaisal committed:
fix: add 1inch and kyberswap client code

felixfaisal committed:
fix: add okx client

felixfaisal committed:
refactor: library structure

felixfaisal committed:
refactor: decouple client code and types

felixfaisal committed:
fix: create evm manageer, construct inch and okx tx

felixfaisal committed:
fix: calculate gas price for all transactions

felixfaisal committed:
fix: add fn to ethereum traits

felixfaisal committed:
fix: construct unsigned txs for swapping

felixfaisal committed:
fix: update client to use endpoints

felixfaisal committed:
refactor: fix warnings for async trait

felixfaisal committed:
fix: add integration tests

felixfaisal committed:
fix: add dev deps

felixfaisal committed:
fix: add approve tx in ethereum client

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

Signed-off-by: Faisal Ahmed  <42486737+felixfaisal@users.noreply.github.com>

felixfaisal committed:
refactor: taplo fmt

felixfaisal committed:
refactor: cargo fmt

felixfaisal committed:
refactor: cargo fmt

kziemianek commented on tee-worker/omni-executor/evm-aggregators-client/src/common.rs:
```diff
@@ -0,0 +1,143 @@
+	pub is_pre_cross: bool,
+}
+
+// TODO: Need to check for Okx native address too
```
Could this todo be resolved ?

felixfaisal replied:
Done ✅

kziemianek commented on tee-worker/omni-executor/evm-aggregators-client/src/inch_client/client.rs:
```diff
@@ -0,0 +1,57 @@
+
+		let status = response.status();
+		let response = response.error_for_status().map_err(|e| {
+			error!("add_wallet request failed with status: {}, error: {:?}", status, e);
```
I think error message should be updated

felixfaisal replied:
Done ✅

kziemianek commented on tee-worker/omni-executor/evm-aggregators-client/src/inch_client/client.rs:
```diff
@@ -0,0 +1,57 @@
+use log::error;
+use reqwest::{Client, Error};
+
+pub const BASIC_ENDPOINT: &str = "https://api.1inch.dev";
```
I would suggest accepting url as constructor param instead of hardcoding it.

felixfaisal replied:
Passing the base_url as constructor param now ✅

kziemianek commented on tee-worker/omni-executor/evm-aggregators-client/src/inch_client/types.rs:
```diff
@@ -0,0 +1,96 @@
+}
+
+impl SwapResponse {
+	pub fn get_transaction_data(self) -> Result<(Vec<u8>, Address, U256, u64), ()> {
```
`tx` is already of type `TransactionData` but this function returns tuple, it's a bit confusing.

felixfaisal replied:
Yeah this is a bit confusing 
So `SwapResponse{tx: Transaction Data...}` -> We then deserialise few fields into `(Vec<u8>, Address, U256 and u64)` 
Ideally, I want to adjust the fields of `TransactionData` such that we can directly serialise into it using serde and we won't have to this additionally handling of conversion of `String` into the types we want. 

However, without having done testing against the aggregators itself, I have to rely on the pumpx implementation, which does it like this, once i can perform testing(P-1590), i can then simplify it accordingly.

kziemianek replied:
Ok, but it's not simple conversion, it also:
* adjusts gas
* makes final_gas capped by GAS_LIMIT

felixfaisal replied:
This can be decoupled from this function too, it is same for all swap responses, we are basically increasing gas by 50%. we can also do it where this function is currently being called.

kziemianek commented on tee-worker/omni-executor/evm-aggregators-client/src/inch_client/types.rs:
```diff
@@ -0,0 +1,96 @@
+}
+
+impl SwapRequest {
+	// TODO, adding protocols is conditional
```
Can this be resolved ?

felixfaisal replied:
Done ✅

kziemianek commented on tee-worker/omni-executor/evm-aggregators-client/src/okx_client/types.rs:
```diff
@@ -0,0 +1,256 @@
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Erc1599Protocol {
```
Is it possible to have basic ser/de tests containing example jsons ?

felixfaisal replied:
So i will be doing this as part of  P-1590, mainly because i can test against aggregators and generate fixtures/jsons which i can later use for unit tests.

felixfaisal committed:
fix: resolve PR comments

felixfaisal committed:
fix: dont use hardcoded urls

felixfaisal committed:
refactor: cargo fmt

felixfaisal committed:
refactor: cargo clippy

felixfaisal committed:
fix: unit tests

felixfaisal committed:
refactor: cargo fmt

kziemianek commented on tee-worker/omni-executor/ethereum-rpc/src/lib.rs:
```diff
@@ -100,6 +102,24 @@ impl RpcProvider for AlloyRpcProvider {
+
+		let pending_transaction_count = provider
+			.get_transaction_count(address)
+			.block_id(BlockId::Number(BlockNumberOrTag::Pending))
```
Are you sure this returns pending transactions count ?
Not overall transaction count (but based on pending - not included - block) ?

So both `transaction_count` and `pending_transaction_count` include already confirmed txs.

felixfaisal replied:
Yeah, you're right, my bad 😅 , I've updated it to only use pending transaction count.

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

kziemianek commented on tee-worker/omni-executor/evm-aggregators-client/src/okx_client/mod.rs:
```diff
@@ -0,0 +1,2 @@
+pub mod client;
```
Please include license headers

felixfaisal replied:
Done ✅

kziemianek commented on tee-worker/omni-executor/evm-aggregators-client/src/evm_tx_manager.rs:
```diff
@@ -0,0 +1,443 @@
+			..Default::default()
+		};
+
+		if is_native_token(&create_market_tx.out_token_ca.into_bytes()) {
```
Does this mean we can swap only from/to native token ?

felixfaisal replied:
Yeah it's based on this - 

https://github.com/dexs-k/dexs-backend/blob/9404cdb153b0a412d198874b2df40fd5d2891779/apps/trade/internal/chain/evm/aggregator.go#L32C47-L32C63 

The comment above it translates to - "Only native coins are supported for now"

kziemianek commented on tee-worker/omni-executor/evm-aggregators-client/src/common.rs:
```diff
@@ -0,0 +1,144 @@
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CreateMarketTx {
```
Unused fields:
  - omni_account: String
  - user_id: u64
  - user_wallet_id: u32
  - is_anti_mev: bool
  - is_auto_slippage: bool
  - out_decimal: u8
  - pair_addr: String
  - price: String
  - use_price_limit: bool
  - in_token_program: String
  - out_token_program: String

Does it make sense to keep them ?

felixfaisal replied:
We don't need them for now, we may need it later, but no point keeping them for now, so removed.

kziemianek commented on tee-worker/omni-executor/evm-aggregators-client/src/kyber_client/client.rs:
```diff
@@ -0,0 +1,97 @@
+			.send()
+			.await
+			.map_err(|e| {
+				error!("Failed to send add_wallet request: {:?}", e);
```
Few error messages should be adjusted.
This for example is not related to add_wallet functionality.

felixfaisal replied:
Updated all messages ✅

kziemianek commented on tee-worker/omni-executor/evm-aggregators-client/src/kyber_client/client.rs:
```diff
@@ -0,0 +1,97 @@
+			e
+		})?;
+		response.json().await.map_err(|e| {
+			error!("Failed to parse 1inch swap response: {:?}", e);
```
This is is kyber not 1inch context

felixfaisal replied:
All error messages updated ✅

kziemianek commented on tee-worker/omni-executor/evm-aggregators-client/src/kyber_client/client.rs:
```diff
@@ -0,0 +1,97 @@
+		})
+	}
+
+	async fn swap(&self, chain_id: u64, swap_request: SwapRequest) -> Result<SwapResponse, Error> {
```
maybe `build_route` ?

felixfaisal replied:
I think it's fine to keep it `swap` even though the endpoint is `build` something, mainly because it ultimately constructs a swap transaction and returns it, and it's consistent with other clients. So I think it's okay.

kziemianek commented on tee-worker/omni-executor/evm-aggregators-client/src/kyber_client/types.rs:
```diff
@@ -0,0 +1,225 @@
+
+impl SwapResponse {
+	pub fn get_transaction_data(self) -> Result<(Vec<u8>, Address, U256, u64), ()> {
+		let data = hex::decode(&self.data).map_err(|e| {
```
Should we inspect `data` to make sure at least that `in` and `out` tokens 'ca` match expectations ?

felixfaisal replied:
I think it's challenging to do that, as the backend solely relies on the aggregator to provide the correct transaction data and we don't know which function call this is to correctly abi decode it. We can create a separate issue for this if we really want to verify the transaction, as this would require some investigation.

kziemianek commented on tee-worker/omni-executor/evm-aggregators-client/src/inch_client/types.rs:
```diff
@@ -0,0 +1,98 @@
+use std::str::FromStr;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SwapResponse {
```
`#[serde(rename_all = "camelCase")]`

should work in general without renaming fields separately

felixfaisal replied:
Updated ✅

Kailai-Wang commented on tee-worker/omni-executor/ethereum-rpc/src/lib.rs:
```diff
@@ -100,6 +102,24 @@ impl RpcProvider for AlloyRpcProvider {
+			.await
+			.map_err(|e| error!("Could not get balance: {:?}", e))?;
+
+		Ok(transaction_count + pending_transaction_count)
```
I don't think that's correct - I assume `get_transaction_count(address, Some(BlockId::Number(BlockNumber::Pending)))` is the only thing you need to calculate the pending nonce, nothing else

felixfaisal replied:
Updated this, my bad 😅 I thought it's a separate counter

Kailai-Wang commented on tee-worker/omni-executor/evm-aggregators-client/src/evm_tx_manager.rs:
```diff
@@ -0,0 +1,443 @@
+		})?;
+
+		let is_buy = is_native_token(&tx.in_token_ca.clone().into_bytes());
+		let platform = Platform::KyberSwap;
```
It seems `platform` is an immutable value - does it mean we always use kyberswap? If yes then why inch and okx client exist? (1:1 refactored from golang?)

felixfaisal replied:
Yea it's 1:1 refactored.

Kailai-Wang reviewed (commented):
Thank you for putting significant efforts on writing and implementing those interfaces - those must be non-trivial work, I appreciate it

Before we going into details of different client impl, I'm wondering are you aware of the heima-router contract:
https://github.com/dexs-k/dexs-contracts/blob/router/src/HeimaRouter.sol

which should be the (maybe only?) entrypoint for swapping: [golang code](https://github.com/dexs-k/dexs-backend/blob/feat/trade-heima/apps/trade/internal/chain/evm/heima.go#L90)

Would that make life easier?

It calls kyberswap under the hood.

Kailai-Wang commented:
As discussed, we could still keep other clients impl, but for the current swap it's preferred to use heima router contract (so that the entrypoint is abstracted and we don't care how it's underlying implemented)

Let's keep it 1:1 to golang impl

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

Signed-off-by: Faisal Ahmed  <42486737+felixfaisal@users.noreply.github.com>

felixfaisal committed:
fix: build

felixfaisal committed:
refactor: add license headers to all files

felixfaisal committed:
refactor: consistent error handling messages

felixfaisal committed:
refactor: use serde rename_all

felixfaisal committed:
refactor: remove unused fields

felixfaisal committed:
fix: use in token from native address check

felixfaisal committed:
fix: use just pending count transaction count as nonce

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

Kailai-Wang commented on tee-worker/omni-executor/evm-aggregators-client/src/common.rs:
```diff
@@ -0,0 +1,149 @@
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+pub const WRAPPED_MAINNET_ETH: &[u8] = b"0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2";
+pub const WRAPPPED_BASE_ETH: &[u8] = b"0x4200000000000000000000000000000000000006";
+pub const GOERLI_WETH: &[u8] = b"0xB4FBF271143F4FBf7B91A5ded31805e42b2208d6";
```
Could we have more consistent naming? Like `WRAPPED_ETH` -> `WETH`? (doesn't matter which one, just to be consistent)

What's even better is to have a `chain_id` -> `address` mapping

felixfaisal replied:
I think this is only used for the function `is_native_token`, i don't think it will be extended.

Kailai-Wang commented on tee-worker/omni-executor/evm-aggregators-client/src/common.rs:
```diff
@@ -0,0 +1,149 @@
+pub const WRAPPPED_BASE_ETH: &[u8] = b"0x4200000000000000000000000000000000000006";
+pub const GOERLI_WETH: &[u8] = b"0xB4FBF271143F4FBf7B91A5ded31805e42b2208d6";
+pub const WRAPPED_BNB: &[u8] = b"0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c";
+pub const NATIVE_TOKEN: &[u8] = b"0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee";
```
Can we be more explicit like `NATIVE_ETH`?

felixfaisal replied:
It's actually a `null` address, just used the same constants as golang impl.

Kailai-Wang commented on tee-worker/omni-executor/evm-aggregators-client/src/common.rs:
```diff
@@ -0,0 +1,149 @@
+	pub is_pre_cross: bool,
+}
+
+pub fn is_native_token(token: &[u8]) -> bool {
```
I this this fn is widely used when creating tx, I don't see where `construct_unsigned_market_tx` is used yet, so I can't really tell what the `token` value would be, but just a reminder:

In RPC params, we use an **empty** `token_ca` address to represent "native token".

felixfaisal replied:
Yeah we need to do some pre-processing for it, I'm unsure however which branch/tag version is deployed to understand it better.

Kailai-Wang commented on tee-worker/omni-executor/evm-aggregators-client/src/evm_tx_manager.rs:
```diff
@@ -0,0 +1,459 @@
+			..Default::default()
+		};
+
+		if is_native_token(&create_market_tx.in_token_ca.into_bytes()) {
+			swap_request.from_token_address = NATIVE_ADDRESS.to_string();
+		} else {
+			swap_request.to_token_address = NATIVE_ADDRESS.to_string();
+		}
```
I don't quite understand this logic - then what is `to_token_address: create_market_tx.out_token_ca.clone()` used for?

felixfaisal replied:
I took this logic as is from the golang implementation, I assume the if the `in_token` is not `NATIVE_TOKEN` then it is assumed that `out_token` is a `NATIVE_TOKEN`, as the comment read they only support to and fro from Native tokens.

Kailai-Wang commented on tee-worker/omni-executor/evm-aggregators-client/src/evm_tx_manager.rs:
```diff
@@ -0,0 +1,459 @@
+
+			let approve_tx = self
+				.eth_client
+				.construct_approve_erc20_tx(
```
Is this approval mandatory? It seems to be executed for every swap, but I don't recall we have such flow in the current impl

felixfaisal replied:
We don't need this for every swap, if it's `BUY` we just check balance, We need this for `SELL` , that is we approve ERC20 on behalf of the aggregators approve contract, so that we can then proceed with a swap transaction. I think that's the general flow of using a swap in EVM chains, we first approve and then swap.

felixfaisal replied:
I think for cross-chain, it's always `BUY`

Kailai-Wang commented on tee-worker/omni-executor/evm-aggregators-client/src/evm_tx_manager.rs:
(Not necessarily in this PR): do you consider extending `SingleChainSwapProvider` (probably more trait def)?

felixfaisal replied:
Yeah I'm extending it in the PR that uses `HeimaRouter.sol`

Kailai-Wang commented:
@kziemianek could you please also take a look if your comments are resolved?

kziemianek commented:
> @kziemianek could you please also take a look if your comments are resolved?

Most of them are. It's fine for me to get it merged in this shape as long as other things are correct. I'm not familiar with golang impl.

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

Kailai-Wang commented:
Additional Q to @wli-pro @0xverin - do we have integration/regression test that can ensure nothing is broken if this PR is merged? 

Not meaning we should fully dedicate it to making one (as swapping isn't the top 1 prior), just asking if we have one already or easy to write one

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

Signed-off-by: Faisal Ahmed  <42486737+felixfaisal@users.noreply.github.com>

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

Kailai-Wang commented:
@gemini-cli /review

Kailai-Wang commented:
@claude review this PR. Be concise. Highlight typos, logic bugs, maintainability issues.

BillyWooo commented:
> **Claude finished @Kailai-Wang's task** —— [View job](https://github.com/litentry/heima/actions/runs/16835977132)
> 
> ### Code Review Findings
> **Typos & Naming Issues:**
> 
> 1. **common.rs:23** - Typo in constant name: `WRAPPPED_BASE_ETH` should be `WRAPPED_BASE_ETH`
> 2. **okx_client/client.rs:91** - Incorrect error message: "Failed to send add_wallet request" should be "Failed to get gas price"
> 3. **okx_client/client.rs:97,101** - Inconsistent error messages: refers to "okx_swap" and "1inch swap" in gas price context
> 
> **Logic Bugs:**
> 
> 1. **evm_tx_manager.rs:105** - Potential panic on HashMap access with `INCH_DEX_IDS_MAP[&chain_id]` - should handle missing chain_id gracefully
> 2. **evm_tx_manager.rs:155,157** - Variable shadowing: `fee_bps` is declared as string but used for percent calculation
> 3. **evm_tx_manager.rs:259** - `.to_i64()` called on u32 without error handling
> 4. **evm_tx_manager.rs:340** - `.unwrap()` used for amount parsing - could panic
> 5. **evm_tx_manager.rs:393** - `.unwrap()` used for address conversion - could panic
> 6. **inch_client/types.rs:48** - Gas adjustment formula `gas.mul(Decimal::new(1, 5))` means multiply by 0.00001, likely intended `Decimal::new(15, 1)` (1.5x)
> 
> **Maintainability Concerns:**
> 
> 1. **common.rs:83-105** - `DECIMALS_TO_VALUE` HashMap could overflow for values >18 decimals (i64 limit)
> 2. **evm_tx_manager.rs:356** - Hardcoded `Platform::KyberSwap` makes other client implementations dead code
> 3. **evm_tx_manager.rs:110-114** - Complex native token logic duplicated across all three construct functions
> 4. Multiple `.map_err(|_|)` patterns lose original error context making debugging difficult
> 5. Extensive use of `String` instead of more type-safe alternatives for addresses/amounts
> 
> **Additional Notes:**
> 
> * Gas adjustment logic is inconsistent across aggregators (all multiply by 0.00001)
> * Error handling relies heavily on logging without propagating structured errors
> * Several TODO comments from review process remain unresolved

@felixfaisal can you please check these feedback? Thanks.

felixfaisal committed:
feat(errors): add comprehensive error handling with ClientError and ClientResult

  - Add thiserror-based error types for better error handling
  - Replace Result<T, ()> with proper ClientResult<T> type
  - Include automatic conversions from common error types (reqwest, serde_json, hex)
  - Add specific error variants for validation, API, and parsing failures

felixfaisal committed:
feat(types): add strongly-typed wrappers for improved type safety

  - Add Address, ChainId, TokenAmount, Slippage, and GasLevel wrappers
  - Include validation and proper parsing with error handling
  - Prevent misuse of primitive types in function parameters
  - Add Display and Debug implementations for better debugging

felixfaisal committed:
feat(transaction): add shared transaction data extraction with gas calculation fix

  - Create TransactionDataExtractor trait to eliminate code duplication
  - Fix critical gas calculation bug: Decimal::new(1,5) → Decimal::new(15,1)
  - Add comprehensive unit tests with mock implementations
  - Apply 1.5x gas multiplier correctly and cap at GAS_LIMIT

felixfaisal committed:
feat(deps): add thiserror and hex dependencies for better error handling

  - Add thiserror = 1.0 for structured error types
  - Add hex = 0.4 for address parsing and validation
  - Update dependencies to support improved type safety

felixfaisal committed:
feat(evm-aggregators): comprehensive refactor with error handling and type safety

  Major improvements based on PR #3540 feedback:

  **Critical Fixes:**
  - Fix gas calculation bug: Decimal::new(1,5) → Decimal::new(15,1) for 1.5x multiplier
  - Replace panic-prone unwrap() calls with proper error handling
  - Fix WRAPPED_BASE_ETH typo and unsafe division operations

  **Architecture Improvements:**
  - Add comprehensive ClientError types with thiserror
  - Create strongly-typed wrappers (Address, ChainId, TokenAmount, etc.)
  - Implement shared TransactionDataExtractor trait (-60 lines duplication)
  - Add builder patterns for all swap request types with validation

  **Testing & Documentation:**
  - Add comprehensive unit tests for gas calculation and slippage conversion
  - Document native token swap logic and basis points conversion
  - Add 14 unit tests with 100% pass rate covering edge cases

  **Performance:**
  - Convert GAS_LIMIT to const and optimize HashMap allocations
  - Reduce memory allocations and improve type safety

felixfaisal commented:
@claude review this PR. Be concise. Highlight typos, logic bugs, maintainability issues.

felixfaisal committed:
refactor: fix error messages

felixfaisal committed:
refactor: use builder pattern for swap request, and consistent error log in okx_client

felixfaisal committed:
fix: update error handling and unit tests

felixfaisal committed:
fix: typo

felixfaisal committed:
Merge branch 'dev' into p-1507-construct-evm-transactions

Signed-off-by: Faisal Ahmed  <42486737+felixfaisal@users.noreply.github.com>

felixfaisal commented:
@claude review this PR. Be concise. Highlight typos, logic bugs, maintainability issues.

felixfaisal commented:
Superseded by #3684
